impl/spanner/mutations/table.go

```diff
@@ -0,0 +1,320 @@
+				return err
+			}
+			logIDs = append(logIDs, logID)
+			return nil
+		}); err != nil {
+		return nil, err
+	}
+	if len(logIDs) == 0 {
+		return nil, status.Errorf(codes.NotFound, "not found")
+	}
+	return logIDs, nil
+}
+
+// Send submits an item to the queue and returns the commit timestamp.
+func (t *Table) Send(ctx context.Context, directoryID string, logID int64, entries ...*pb.EntryUpdate) (water.Mark, error) {
+	if len(entries) == 0 {
+		return water.Mark{}, fmt.Errorf("no entries to send")
+	}
+	type Cols struct {
+		DirectoryID string
+		LogID       int64
+		Timestamp   time.Time
+		LocalID     int64
+		Mutation    []byte
+	}
+	ms := make([]*spanner.Mutation, 0, len(entries))
+	for i, e := range entries {
+		mBytes, err := proto.Marshal(e)
+		if err != nil {
+			return water.Mark{}, err
+		}
+		m, err := spanner.InsertStruct(mutTable, Cols{
+			DirectoryID: directoryID,
+			LogID:       logID,
+			Timestamp:   spanner.CommitTimestamp,
+			LocalID:     int64(i),
+			Mutation:    mBytes,
+		})
+		if err != nil {
+			return water.Mark{}, err
+		}
+		ms = append(ms, m)
+	}
+
+	commitTimestamp, err := t.client.ReadWriteTransaction(ctx,
+		func(_ context.Context, txn *spanner.ReadWriteTransaction) error {
+			return txn.BufferWrite(ms)
+		})
+	if err != nil {
+		return water.Mark{}, err
+	}
+	return timeToMark(commitTimestamp), nil
+}
+
+// HighWatermark returns the number of items and the highest timestamp
+// (exclusive) up to batchSize items after start (inclusive).
+// Because timestamps can group more than one item, count may slightly exceed batchSize.
+func (t *Table) HighWatermark(ctx context.Context, directoryID string, logID int64, start water.Mark, batchSize int32) (count int32, high water.Mark, err error) {
+	// TODO: Replace with MAX(Timestamp) when spansql supports aggregate operators.
+	stmt := spanner.NewStatement(`
+	SELECT
+	   Timestamp
+	FROM
+	   Mutations
+	WHERE
+	   DirectoryID = @directoryID AND
+	   LogID = @logID AND
+	   Timestamp >= @start
+	ORDER BY Timestamp
+	LIMIT @limit`)
+	stmt.Params["directoryID"] = directoryID
+	stmt.Params["logID"] = logID
+	stmt.Params["start"] = markToTime(start)
+	stmt.Params["limit"] = int64(batchSize)
+
+	var cnt int64
+	var max time.Time
+	rtx := t.client.Single().WithTimestampBound(spanner.MaxStaleness(readStaleness))
+	defer rtx.Close()
+	if err := rtx.Query(ctx, stmt).Do(
+		func(row *spanner.Row) error {
+			cnt++
+			var watermark time.Time
+			if err := row.Column(0, &watermark); err != nil {
+				return err
+			}
+			if watermark.After(max) {
+				max = watermark
+			}
+			return nil
+		}); err != nil {
+		return 0, water.Mark{}, err
+	}
+	if cnt == 0 {
+		// When there are no rows, return the start time as the highest timestamp.
+		return 0, start, nil
+	}
+	return int32(cnt), timeToMark(max).Add(1), nil
+}
+
+// ReadLog reads all mutations between [low, high).
+// If limit is used, ReadLog ensures that complete batches are returned, which could slightly exceed limit.
```

Comment:
```suggestion
// If limit is used, ReadLog ensures that complete batches are returned, which could exceed limit.
```

From the perspective of this library, there is no control over how big or small this could be.

May be worth adding clarification that a batch is all entries with the same timestamp, because that wasn't obvious to me.